fe/fe-core/src/main/java/org/apache/doris/common/cache/NereidsSqlCacheManager.java

```diff
@@ -350,13 +346,35 @@ private boolean tablesOrDataChanged(Env env, SqlCacheContext sqlCacheContext) {
                 return true;
             }
             OlapTable olapTable = (OlapTable) tableIf;
-            Collection<Long> partitionIds = scanTable.getScanPartitions();
-            olapTable.getVersionInBatchForCloudMode(partitionIds);
-
-            for (Long scanPartitionId : scanTable.getScanPartitions()) {
-                Partition partition = olapTable.getPartition(scanPartitionId);
+            List<Column> currentFullSchema = olapTable.getFullSchema();
+            List<SqlCacheContext.ColumnSection> cacheFullSchema = scanTable.latestColumns;
+            if (currentFullSchema.size() != cacheFullSchema.size()) {
+                return true;
+            }
+            for (int i = 0; i < currentFullSchema.size(); i++) {
+                Column currentColumn = currentFullSchema.get(i);
+                SqlCacheContext.ColumnSection cacheColumn = cacheFullSchema.get(i);
+                if ((currentColumn == null || cacheColumn == null)
+                    || (currentColumn.hashCode() != cacheColumn.getHashCode())) {
+                    return true;
+                }
+                if (!Objects.equals(currentColumn.getName(), cacheColumn.getName())) {
+                    return true;
+                }
+                if (!Objects.equals(currentColumn.getType().toString(), cacheColumn.getType())) {
+                    return true;
+                }
+            }
```

Comment:
1. You should move to the previous loop of `sqlCacheContext.getUsedTables()` because
```
        // the query maybe scan empty partition of the table, we should check these table version too,
        // but the table not exists in sqlCacheContext.getScanTables(), so we need check here.
        // check table type and version
```

2. If we already check the columns for table version `x`, can we skip check columns next time if the table version is `x`?